compiler/rustc_lint/src/types.rs

```diff
@@ -1550,62 +1540,56 @@ impl InvalidAtomicOrdering {
-                    cx.struct_span_lint(INVALID_ATOMIC_ORDERING, failure_order_arg.span, |diag| {
-                        let msg = format!(
-                            "{}'s failure ordering may not be stronger than the success ordering of `{}`",
-                            method,
-                            success_ord,
-                        );
-                        diag.build(&msg)
-                            .help(&format!("consider using {} instead", suggested))
-                            .emit();
-                    });
-                }
-            }
+        let Some((method, args)) = Self::inherent_atomic_method_call(cx, expr, &[sym::fetch_update, sym::compare_exchange, sym::compare_exchange_weak])
+            else {return };
+
+        let (success_order_arg, fail_order_arg) = match method {
+            sym::fetch_update => (&args[1], &args[2]),
+            sym::compare_exchange | sym::compare_exchange_weak => (&args[3], &args[4]),
+            _ => return,
+        };
+
+        let Some(fail_ordering) = Self::match_ordering(cx, fail_order_arg) else { return };
+
+        if matches!(fail_ordering, sym::Release | sym::AcqRel) {
+            cx.struct_span_lint(INVALID_ATOMIC_ORDERING, fail_order_arg.span, |diag| {
+                diag.build(&format!(
+                    "{method}'s failure ordering may not be `Release` or `AcqRel`, \
+                    since a failed {method} does not result in a write",
+                ))
+                .span_label(fail_order_arg.span, "invalid failure ordering")
+                .help("consider using Acquire or Relaxed failure ordering instead")
+                .emit();
+            });
+        }
+
+        let Some(success_ordering) = Self::match_ordering(cx, success_order_arg) else { return };
+
+        if matches!(
+            (success_ordering, fail_ordering),
+            (sym::Relaxed | sym::Release, sym::Acquire)
+                | (sym::Relaxed | sym::Release | sym::Acquire | sym::AcqRel, sym::SeqCst)
+        ) {
+            let success_suggestion =
+                if success_ordering == sym::Release && fail_ordering == sym::Acquire {
+                    sym::AcqRel
+                } else {
+                    fail_ordering
+                };
+            cx.struct_span_lint(INVALID_ATOMIC_ORDERING, success_order_arg.span, |diag| {
+                diag.build(&format!(
+                    "{method}'s success ordering must be at least as strong as its failure ordering"
+                ))
+                .span_label(fail_order_arg.span, format!("{fail_ordering} failure ordering"))
+                .span_label(success_order_arg.span, format!("{success_ordering} success ordering"))
+                .span_suggestion_short(
+                    success_order_arg.span,
+                    format!("consider using {success_suggestion} success ordering instead"),
+                    success_suggestion.to_string(),
```

Comment:
Somewhat concerned about the applicability of the suggestion because we don't use the full path here, only the variant name. Maybe formatting this to be

```suggestion
                    format!("std::sync::atomic::Ordering::{success_suggestion}"),
```

so that it's always applicable might be better (even if it results in uglier code).

Reply:
Wouldn't hard-coding `std` break for `#![no_std]` crates?

Reply:
It would, but we can also assume that practitioners writing `no_std` code can figure it out. Ideally we would check if we have `std` available :)